openedx/core/djangoapps/auth_exchange/forms.py

```diff
@@ -2,33 +2,113 @@
+from django.conf import settings
+from django.utils.encoding import smart_text
+from django.utils.translation import ugettext as _
 from oauth2_provider.models import Application
-from provider.forms import OAuthForm, OAuthValidationError
-from provider.oauth2.forms import ScopeChoiceField, ScopeMixin
-from provider.oauth2.models import Client
 from requests import HTTPError
 from social_core.backends import oauth as social_oauth
 from social_core.exceptions import AuthException
 
 from third_party_auth import pipeline
 
 
-class AccessTokenExchangeForm(ScopeMixin, OAuthForm):
+class OAuthValidationError(Exception):
+    """
+    Exception to throw inside :class:`AccessTokenExchangeForm` if any OAuth2 related errors
+    are encountered such as invalid grant type, invalid client, etc.
+    :attr:`OAuthValidationError` expects a dictionary outlining the OAuth error
+    as its first argument when instantiating.
+    :example:
+    ::
+        class GrantValidationForm(AccessTokenExchangeForm):
+            grant_type = forms.CharField()
+            def clean_grant(self):
+                if not self.cleaned_data.get('grant_type') == 'code':
+                    raise OAuthValidationError({
+                        'error': 'invalid_grant',
+                        'error_description': "%s is not a valid grant type" % (
+                            self.cleaned_data.get('grant_type'))
+                    })
+    """
+
+
+class ScopeChoiceField(forms.ChoiceField):
+    """
+    Custom form field that seperates values on space
+    """
+    widget = forms.SelectMultiple
+
+    def to_python(self, value):
+        if not value:
+            return []
+
+        # value may come in as a string.
+        # try to parse and
+        # ultimately return an empty list if nothing remains -- this will
+        # eventually raise an `OAuthValidationError` in `validate` where
+        # it should be anyways.
+        if not isinstance(value, (list, tuple)):
+            value = value.split(' ')
+
+        # Split values into list
+        return u' '.join([smart_text(val) for val in value]).split(u' ')
+
+    def validate(self, value):
+        """
+        Validates that the input is a list or tuple.
+        """
+        if self.required and not value:
+            raise OAuthValidationError({'error': 'invalid_request'})
+
+        # Validate that each value in the value list is in self.choices.
+        for val in value:
+            if not self.valid_value(val):
+                raise OAuthValidationError({
+                    'error': 'invalid_request',
+                    'error_description': _("'%s' is not a valid scope.") %
+                    val})
+
+
+class AccessTokenExchangeForm(forms.Form):
     """Form for access token exchange endpoint"""
+
     access_token = CharField(required=False)
-    scope = ScopeChoiceField(choices=SCOPE_NAMES, required=False)
+    OAUTH2_PROVIDER_setting = getattr(settings, "OAUTH2_PROVIDER", {})
```

Comment:
This variable name is a bit odd. I would just call it `OAUTH2_PROVIDER` because it's only going to be scoped to this class anyway.

Reply:
changed name